src/libopensc/card-edo.c

```diff
@@ -0,0 +1,299 @@
+}
+
+
+static int edo_select_name(struct sc_card* card, const u8* name, size_t namelen, struct edo_buff* buff) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct sc_apdu apdu;
+	sc_format_apdu_ex(&apdu, 00, 0xA4, 0x04, 0x00, name, namelen, buff->val, sizeof buff->val);
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "SW check failed");
+	buff->len = apdu.resplen;
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+static int edo_select_path(struct sc_card* card, const u8* path, size_t pathlen, struct edo_buff* buff) {
+	LOG_FUNC_CALLED(card->ctx);
+	while (pathlen >= 2) {
+		if (path[0] == 0x3F && path[1]  == 0x00)
+			LOG_TEST_RET(card->ctx, edo_select_mf(card, buff), "MF select failed");
+		else if (path[0] == 0xDF)
+			LOG_TEST_RET(card->ctx, edo_select_df(card, path, buff), "DF select failed");
+		else if (pathlen == 2)
+			LOG_TEST_RET(card->ctx, edo_select_ef(card, path, buff), "EF select failed");
+		else
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+		path += 2;
+		pathlen -= 2;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Selects file specified by given path.
+ *
+ * Card does not support selecting file at once, that's why it have to be done in following way:
+ * 1. Select AID if provided,
+ * 2. Select MF if provided,
+ * 3. Select DF until provided,
+ * 4. Select EF if provided.
+ */
+static int edo_select_file(struct sc_card* card, const struct sc_path* in_path, struct sc_file** file_out) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct edo_buff buff;
+
+	switch (in_path->type) {
+		case SC_PATH_TYPE_PATH:
+		case SC_PATH_TYPE_FILE_ID:
+			if (in_path->aid.len)
+				LOG_TEST_RET(card->ctx, edo_select_name(card, in_path->aid.value, in_path->aid.len, &buff), "Select AID failed");
+			if (in_path->len)
+				LOG_TEST_RET(card->ctx, edo_select_path(card, in_path->value, in_path->len, &buff), "Select path failed");
+			break;
+		case SC_PATH_TYPE_DF_NAME:
+			LOG_TEST_RET(card->ctx, edo_select_name(card, in_path->value, in_path->len, &buff), "Select AID failed");
+			break;
+		default:
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	}
+
+	if (file_out) {
+		if (buff.len < 2)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_UNKNOWN_DATA_RECEIVED);
+		if (!(*file_out = sc_file_new()))
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		(*file_out)->path = *in_path;
+		LOG_TEST_RET(card->ctx, card->ops->process_fci(card, *file_out, buff.val, buff.len), "Process FCI failed");
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Computes ECDSA signature.
+ *
+ * If ECDSA was used, the ASN.1 sequence of integers R,S returned by the
+ * card needs to be converted to the raw concatenation of R,S for PKCS#11.
+ */
+static int edo_compute_signature(struct sc_card* card, const u8* data, size_t datalen, u8* out, size_t outlen) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct edo_buff buff = {{}, sizeof buff.val};
```

Comment:
For readability, you should use `u8 val[SC_MAX_APDU_RESP_SIZE];` directly here.

Reply:
IMO it is good to contextually tie buffer with it length

Reply:
I thought length is required there, I see it can be simply replaced with plain buffer